## Housing Rental Analysis for San Francisco

In [18]:
# Import the required libraries and dependencies
import numpy as np
import pandas as pd
import hvplot.pandas
from pathlib import Path
#import holoviews as hv
import os
# from bokeh.models import HoverTool

In [2]:
# Using the read_csv function and Path module, create a DataFrame 
# by importing the sfo_neighborhoods_census_data.csv file from the Resources folder
sfo_data_df = pd.read_csv(
    Path('./Resources/sfo_neighborhoods_census_data.csv')
        )

# Set year as the index
sfo_data_df = sfo_data_df.set_index('year')

# Review the first and last five rows of the DataFrame
display(sfo_data_df.head())
display(sfo_data_df.tail())
sfo_data_df.dtypes


neighborhood  sale_price_sqr_foot  housing_units  gross_rent
year                                                                  
2010      Alamo Square           291.182945         372560        1239
2010        Anza Vista           267.932583         372560        1239
2010           Bayview           170.098665         372560        1239
2010  Buena Vista Park           347.394919         372560        1239
2010  Central Richmond           319.027623         372560        1239

neighborhood  sale_price_sqr_foot  housing_units  gross_rent
year                                                                        
2016          Telegraph Hill           903.049771         384242        4390
2016              Twin Peaks           970.085470         384242        4390
2016  Van Ness/ Civic Center           552.602567         384242        4390
2016       Visitacion Valley           328.319007         384242        4390
2016           Westwood Park           631.195426         384242        4390

neighborhood            object
sale_price_sqr_foot    float64
housing_units            int64
gross_rent               int64
dtype: object

In [3]:
# Create a numerical aggregation that groups the data by the year and then averages the results.
housing_units_by_year = sfo_data_df.groupby('year').mean()

# Review the DataFrame
display(housing_units_by_year.head())

sale_price_sqr_foot  housing_units  gross_rent
year                                                
2010           369.344353       372560.0      1239.0
2011           341.903429       374507.0      1530.0
2012           399.389968       376454.0      2324.0
2013           483.600304       378401.0      2971.0
2014           556.277273       380348.0      3528.0

In [4]:
# Create a visual aggregation explore the housing units by year
housing_units_by_year.hvplot.bar(    
title = "San Francisco Housing Units by Year",
x="year",
y="housing_units",
xlabel="Year",
ylabel="Housing Units",
ylim=(365000,385000),
rot = 0
).opts(
    yformatter='%.0f',
    color="blue",
    hover_color="green",
)

:Bars   [year]   (housing_units)

**Question:** What is the overall trend in housing_units over the period being analyzed?

**Answer:** Total number of housing units is steadily increasing each year. Additionaly the sale price by square footage is increasing and the gross rents have almost tripled during the analysis period.

In [5]:
# Create a numerical aggregation by grouping the data by year and averaging the results
prices_square_foot_by_year = sfo_data_df.groupby('year').mean()

# Review the resulting DataFrame
prices_square_foot_by_year

sale_price_sqr_foot  housing_units  gross_rent
year                                                
2010           369.344353       372560.0      1239.0
2011           341.903429       374507.0      1530.0
2012           399.389968       376454.0      2324.0
2013           483.600304       378401.0      2971.0
2014           556.277273       380348.0      3528.0
2015           632.540352       382295.0      3739.0
2016           697.643709       384242.0      4390.0

**Question:** What is the lowest gross rent reported for the years included in the DataFrame?

**Answer:** The lowest gross rent is $1,239 in 2010

In [6]:
# Filter out the housing_units column, creating a new DataFrame 
# Keep only sale_price_sqr_foot and gross_rent averages per year
prices_square_foot_by_year_filter = prices_square_foot_by_year[["sale_price_sqr_foot","gross_rent"]]

# Review the DataFrame
prices_square_foot_by_year_filter

sale_price_sqr_foot  gross_rent
year                                 
2010           369.344353      1239.0
2011           341.903429      1530.0
2012           399.389968      2324.0
2013           483.600304      2971.0
2014           556.277273      3528.0
2015           632.540352      3739.0
2016           697.643709      4390.0

In [7]:
# Plot prices_square_foot_by_year. 
# Inclued labels for the x- and y-axes, and a title.
prices_square_foot_by_year_filter.hvplot.line(title = "Sale Price per Sq Ft and Avg Gross Rent",
xlabel="Year",
ylabel="Sale price per sq ft")

:NdOverlay   [Variable]
   :Curve   [year]   (value)

**Question:** Did any year experience a drop in the average sale price per square foot compared to the previous year?

**Answer:** Yes there was a slight drop in the average sale price per square foot from 2010 (369) vs 2011 (341)

Question:** If so, did the gross rent increase or decrease during that year?

**Answer:** No the gross rent actually increased during this time.

For this part of the assignment, use interactive visualizations and widgets to explore the average sale price per square foot by neighborhood. To do so, complete the following steps:

Create a new DataFrame that groups the original DataFrame by year and neighborhood. Aggregate the results by the mean of the groups.

Filter out the “housing_units” column to create a DataFrame that includes only the sale_price_sqr_foot and gross_rent averages per year.

Create an interactive line plot with hvPlot that visualizes both sale_price_sqr_foot and gross_rent. Set the x-axis parameter to the year (x="year"). Use the groupby parameter to create an interactive widget for neighborhood.

Style and format the line plot to ensure a professionally styled visualization.

Note that your resulting plot should appear similar to the following image:

In [8]:
# Group by year and neighborhood and then create a new dataframe of the mean values
prices_by_year_by_neighborhood = sfo_data_df.groupby(['year','neighborhood']).mean()

# Review the DataFrame
prices_by_year_by_neighborhood

sale_price_sqr_foot  housing_units  gross_rent
year neighborhood                                                          
2010 Alamo Square                     291.182945       372560.0      1239.0
     Anza Vista                       267.932583       372560.0      1239.0
     Bayview                          170.098665       372560.0      1239.0
     Buena Vista Park                 347.394919       372560.0      1239.0
     Central Richmond                 319.027623       372560.0      1239.0
...                                          ...            ...         ...
2016 Telegraph Hill                   903.049771       384242.0      4390.0
     Twin Peaks                       970.085470       384242.0      4390.0
     Van Ness/ Civic Center           552.602567       384242.0      4390.0
     Visitacion Valley                328.319007       384242.0      4390.0
     Westwood Park                    631.195426       384242.0      4390.0

[397 rows x 3 columns]

In [9]:
# Filter out the housing_units
prices_by_year_by_neighborhood_filter = prices_by_year_by_neighborhood.drop(columns=["housing_units"])

# Review the first and last five rows of the DataFrame
display(prices_by_year_by_neighborhood_filter.head())
display(prices_by_year_by_neighborhood_filter.tail())

sale_price_sqr_foot  gross_rent
year neighborhood                                     
2010 Alamo Square               291.182945      1239.0
     Anza Vista                 267.932583      1239.0
     Bayview                    170.098665      1239.0
     Buena Vista Park           347.394919      1239.0
     Central Richmond           319.027623      1239.0

sale_price_sqr_foot  gross_rent
year neighborhood                                           
2016 Telegraph Hill                   903.049771      4390.0
     Twin Peaks                       970.085470      4390.0
     Van Ness/ Civic Center           552.602567      4390.0
     Visitacion Valley                328.319007      4390.0
     Westwood Park                    631.195426      4390.0

In [10]:
# Use hvplot to create an interactive line plot of the average price per square foot
# The plot should have a dropdown selector for the neighborhood
prices_by_year_by_neighborhood_filter.hvplot.line(
    groupby="neighborhood",
    title="Sale Price Per Square Foot and Avg Gross Rent- by Neighborhood",
    ylabel="Gross Rent/Sale Price Per Square Foot",
    xlabel="Year"
    
)

:DynamicMap   [neighborhood]
   :NdOverlay   [Variable]
      :Curve   [year]   (value)

**Question:** For the Anza Vista neighborhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012? 

**Answer:** # For 2016 the average sale price per square foot for Anza Vista (88) is less than in 2012 (344)

## Build an Interactive Neighborhood Map

For this part of the assignment, explore the geospatial relationships in the data by using interactive visualizations with hvPlot and GeoViews. To build your map, use the `sfo_data_df` DataFrame (created during the initial import), which includes the neighborhood location data with the average prices. To do all this, complete the following steps:

1. Read the `neighborhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighborhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighborhood”.

2. Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighborhood_info_df` that groups the data by neighborhood. Aggregate the results by the `mean` of the group.

3. Review the two code cells that concatenate the `neighborhood_locations_df` DataFrame with the `all_neighborhood_info_df` DataFrame. Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighborhoods_df`. The second cell cleans the data and sets the “Neighborhood” column. Be sure to run these cells to create the `all_neighborhoods_df` DataFrame, which you’ll need to create the geospatial visualization.

4. Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighborhoods_df` DataFrame. Be sure to do the following:

    * Set the `size` parameter to “sale_price_sqr_foot”.

    * Set the `color` parameter to “gross_rent”.

    * Set the `size_max` parameter to “25”.

    * Set the `zoom` parameter to “11”.

5. Use the interactive map to answer the following question:

    * Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?

In [11]:
# Load neighborhoods coordinates data
neighborhood_locations_df = pd.read_csv(
    Path('./Resources/neighborhoods_coordinates.csv')
        )
neighborhood_locations_df=neighborhood_locations_df.set_index('Neighborhood')
# Review the DataFrame
neighborhood_locations_df

Lat         Lon
Neighborhood                             
Alamo Square        37.791012 -122.402100
Anza Vista          37.779598 -122.443451
Bayview             37.734670 -122.401060
Bayview Heights     37.728740 -122.410980
Bernal Heights      37.728630 -122.443050
...                       ...         ...
West Portal         37.740260 -122.463880
Western Addition    37.792980 -122.435790
Westwood Highlands  37.734700 -122.456854
Westwood Park       37.734150 -122.457000
Yerba Buena         37.792980 -122.396360

[73 rows x 2 columns]

In [12]:
# Calculate the mean values for each neighborhood
all_neighborhood_info_df = sfo_data_df.groupby(['neighborhood']).mean()

# Review the resulting DataFrame
all_neighborhood_info_df

sale_price_sqr_foot  housing_units   gross_rent
neighborhood                                                       
Alamo Square                 366.020712      378401.00  2817.285714
Anza Vista                   373.382198      379050.00  3031.833333
Bayview                      204.588623      376454.00  2318.400000
Bayview Heights              590.792839      382295.00  3739.000000
Bernal Heights               576.746488      379374.50  3080.333333
...                                 ...            ...          ...
West Portal                  498.488485      376940.75  2515.500000
Western Addition             307.562201      377427.50  2555.166667
Westwood Highlands           533.703935      376454.00  2250.500000
Westwood Park                687.087575      382295.00  3959.000000
Yerba Buena                  576.709848      377427.50  2555.166667

[73 rows x 3 columns]

In [13]:
# Using the Pandas `concat` function, join the 
# neighborhood_locations_df and the all_neighborhood_info_df DataFrame
# The axis of the concatenation is "columns".
# The concat function will automatially combine columns with
# identical information, while keeping the additional columns.
all_neighborhoods_df = pd.concat(
    [neighborhood_locations_df, all_neighborhood_info_df], 
    axis="columns",
    sort=False
)

# Review the resulting DataFrame
display(all_neighborhoods_df.head())
display(all_neighborhoods_df.tail())

Lat         Lon  sale_price_sqr_foot  housing_units  \
Alamo Square     37.791012 -122.402100           366.020712       378401.0   
Anza Vista       37.779598 -122.443451           373.382198       379050.0   
Bayview          37.734670 -122.401060           204.588623       376454.0   
Bayview Heights  37.728740 -122.410980           590.792839       382295.0   
Bernal Heights   37.728630 -122.443050                  NaN            NaN   

                  gross_rent  
Alamo Square     2817.285714  
Anza Vista       3031.833333  
Bayview          2318.400000  
Bayview Heights  3739.000000  
Bernal Heights           NaN

Lat        Lon  sale_price_sqr_foot  housing_units  \
Yerba Buena      37.79298 -122.39636           576.709848       377427.5   
Bernal Heights        NaN        NaN           576.746488       379374.5   
Downtown              NaN        NaN           391.434378       378401.0   
Ingleside             NaN        NaN           367.895144       377427.5   
Outer Richmond        NaN        NaN           473.900773       378401.0   

                  gross_rent  
Yerba Buena      2555.166667  
Bernal Heights   3080.333333  
Downtown         2817.285714  
Ingleside        2509.000000  
Outer Richmond   2817.285714

In [14]:
# Call the dropna function to remove any neighborhoods that do not have data
all_neighborhoods_df = all_neighborhoods_df.reset_index().dropna()

# Rename the "index" column as "Neighborhood" for use in the Visualization
all_neighborhoods_df = all_neighborhoods_df.rename(columns={"index": "Neighborhood"})

# all_neighborhoods_df = all_neighborhoods_df.sort_values("gross_rent")
#all_neighborhoods_df = all_neighborhoods_df.sort_values("sale_price_sqr_foot")

# Review the resulting DataFrame
display(all_neighborhoods_df.head())
display(all_neighborhoods_df.tail())

Neighborhood        Lat         Lon  sale_price_sqr_foot  \
0      Alamo Square  37.791012 -122.402100           366.020712   
1        Anza Vista  37.779598 -122.443451           373.382198   
2           Bayview  37.734670 -122.401060           204.588623   
3   Bayview Heights  37.728740 -122.410980           590.792839   
5  Buena Vista Park  37.768160 -122.439330           452.680591   

   housing_units   gross_rent  
0       378401.0  2817.285714  
1       379050.0  3031.833333  
2       376454.0  2318.400000  
3       382295.0  3739.000000  
5       378076.5  2698.833333

Neighborhood       Lat         Lon  sale_price_sqr_foot  \
68         West Portal  37.74026 -122.463880           498.488485   
69    Western Addition  37.79298 -122.435790           307.562201   
70  Westwood Highlands  37.73470 -122.456854           533.703935   
71       Westwood Park  37.73415 -122.457000           687.087575   
72         Yerba Buena  37.79298 -122.396360           576.709848   

    housing_units   gross_rent  
68      376940.75  2515.500000  
69      377427.50  2555.166667  
70      376454.00  2250.500000  
71      382295.00  3959.000000  
72      377427.50  2555.166667

In [15]:
# Create a plot to analyze neighborhood info
all_neighborhoods_df.hvplot.points(
    'Lon', 
    'Lat', 
    geo=True,
    size='sale_price_sqr_foot',
    color='gross_rent',
    #size_max=25,
    #zoom=11,
    alpha=0.8,
    tiles='OSM',
    frame_width = 700,
    frame_height = 500,
    hover_cols = "Neighborhood"
)

:Overlay
   .Tiles.I  :Tiles   [x,y]
   .Points.I :Points   [Lon,Lat]   (gross_rent,sale_price_sqr_foot,Neighborhood)

**Question:** Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?

**Answer:** Westward Park has the highest gross rent ($3959). The Union Square District has the highest sale price per square foot ($903)

## Compose Your Data Story

Based on the visualizations that you have created, compose a data story that synthesizes your analysis by answering the following questions:

**Question:**  How does the trend in rental income growth compare to the trend in sales prices? Does this same trend hold true for all the neighborhoods across San Francisco?

**Answer:** # The overall trend is that rents are decreasing and sales prices are increasing although this is not true in all neighborhoods.

**Question:** What insights can you share with your company about the potential one-click, buy-and-rent strategy that they're pursuing? Do neighborhoods exist that you would suggest for investment, and why?

**Answer:** # YOUR ANSWER HERE

In most neighborhood rents are decreasing at the same time as its become more expensive to buy rental properties. The are some exceptions, for example in the Outer Mission neighborhood rents have been rising as sale prices have decreased. In conclusion the tool is very effective for identifying investment opportunities.